### DISCUSSION TITLE
How do I get Pytest to look in my project?

### DISCUSSION THREAD

**Discussion by transcental** [pytest-dev/pytest#9693] [Q&A] [answered]
By default, pytest looks in the `site-packages` folder. This is incredibly annoying as I don't want to upload my package to PyPi just to test it.

Here's my directory tree and the output of `pytest -v`.

```
.
|____CODE_OF_CONDUCT.md
|____pygitapi
| |______init__.py
| |____utils.py
| |____github.py
|____LICENSE
| |____README.md
| |____.gitignore
|____pyproject.toml
|____tests
| |____test_github.py
|____test.py
|____README.md
|____setup.py
|____.gitignore
|____.env
|____.github
| |____workflows
| | |____pytest.yml
| | |____codeql-analysis.yml
| | |____pypi-upload.yml
|____setup.cfg
|____.all-contributorsrc
|____SECURITY.md
```

```zsh
dillonbarnes at Downstairs-iMac in ~/Documents/Coding/GitStats on master!
± pytest -v
======================================================================================== test session starts =========================================================================================
platform darwin -- Python 3.10.0, pytest-7.0.1, pluggy-1.0.0 -- /Library/Frameworks/Python.framework/Versions/3.10/bin/python3.10
cachedir: .pytest_cache
rootdir: /Users/dillonbarnes/Documents/Coding/GitStats
collected 0 items / 1 error                                                                                                                                                                          

=============================================================================================== ERRORS ===============================================================================================
_______________________________________________________________________________ ERROR collecting tests/test_github.py ________________________________________________________________________________
ImportError while importing test module '/Users/dillonbarnes/Documents/Coding/GitStats/tests/test_github.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_github.py:2: in <module>
    from pygitapi import HubAPI
E   ModuleNotFoundError: No module named 'pygitapi'
====================================================================================== short test summary info =======================================================================================
ERROR tests/test_github.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
========================================================================================== 1 error in 0.11s ==========================================================================================
```

Is this a bug or just something I'm doing wrong?

**Comment by nicoddemus** [comment]
Hi @DillonB07,

See https://docs.pytest.org/en/stable/reference/reference.html#confval-norecursedirs, specially:

> Additionally, pytest will attempt to intelligently identify and ignore a virtualenv by the presence of an activation script. Any directory deemed to be the root of a virtual environment will not be considered during test collection unless ‑‑collect‑in‑virtualenv is given. Note also that norecursedirs takes precedence over ‑‑collect‑in‑virtualenv; e.g. if you intend to run tests in a virtualenv with a base directory that matches '.*' you must override norecursedirs in addition to using the ‑‑collect‑in‑virtualenv flag.

So not sure why this detection is not working, but regardless you can configure [`testpaths`](https://docs.pytest.org/en/stable/reference/reference.html#confval-testpaths).

Hope this helps.

**Reply by transcental** [reply]
Thanks, but you misunderstood the issue. Pytest found my tests, but I was testing a package I was developing. The issue was that it was looking for the package in `site-packages` rather than the project directory.

**Reply by nicoddemus** [reply]
I don't think I did, you said:

> The issue was that it was looking for the package in site-packages rather than the project directory.

And I pointed out:

> Additionally, pytest will attempt to intelligently identify and **ignore** a virtualenv by the presence of an activation script.

(Emphasis on **ignore**)

That's why I said I wasn't sure the virtual env detection was not working (and hence pytest ignoring tests on it, as it should), and pointed you to the `testpaths` configuration variable.

**Reply by transcental** [reply]
Sorry, I misunderstood the log. 

It was just a location issue as jwodder mentioned.

**Comment by jwodder** [comment, marked as answer]
(Disregard what I wrote earlier; I misunderstood your problem.)

So your problem is that your tests try to import your project source, but it's not found.  This isn't a pytest thing; it's a basic Python packaging/importing thing.  In order for your package to be importable, it needs to be in `sys.path`.  When you run just `pytest`, the current directory will not be in `sys.path`, but it will be if you instead run `python -m pytest`.  (At least, I'm 99% sure that's how it works; I can't find where this is mentioned in the pytest documentation at the moment.)  So the simple solution is to run `python -m pytest` instead of just `pytest`.

However!  This is not good packaging/testing practice.  What you should do instead is install your package into a virtualenv that also contains pytest, and run pytest inside that venv.  This can be done with [tox](https://tox.wiki/en/latest/) or its cousin [nox](https://nox.thea.codes).  Moreover, you should move your package code (not your tests!) inside a `src/` directory in order to ensure that your tests run against the installed version of your package rather than the local copy, which will help you detect errors in packaging; see [here](https://blog.ionelmc.ro/2014/05/25/python-packaging/#the-structure) and [here](https://hynek.me/articles/testing-packaging/) for more information.

**Reply by transcental** [reply]
Thanks!

I did originally have my code in a `src/` directory, but took it out to try and make `pytest` work.

Tox did the job perfectly. Thank you!

However, I have a slight issue where even if I pass through my env variable to the GitHubaction tox doesn't pick it up? Is there a solution to this that you know of?

**Reply by jwodder** [reply]
Tox purges most environment variables in order to produce a consistent environment.  You can tell it to preserve a given envvar by setting `passenv = NAME_OF_ENVVAR` in your `[testenv]` section.

**Reply by transcental** [reply]
Thanks. That fixed it!
